verifier/src/qe_report_body.rs

```diff
@@ -0,0 +1,550 @@
+///
+/// Preforms step 4 of
+/// <https://api.portal.trustedservices.intel.com/documentation#pcs-enclave-identity-v4>
+///
+/// > 4. Determine a TCB status of the Enclave:
+/// >       a. Retrieve a collection of TCB Levels (sorted by ISVSVNs) from
+/// >          tcbLevels field in Enclave Identity structure.
+/// >       b. Go over the list of TCB Levels (descending order) and find the
+/// >          one that has ISVSVN that is lower or equal to the ISVSVN value
+/// >          from SGX Enclave Report.
+/// >       c. If a TCB level is found, read its status from tcbStatus field,
+/// >          otherwise your TCB Level is not supported.
+#[derive(Debug, Clone)]
+struct QeIsvSvnVerifier {
+    tcb_levels: Vec<TcbLevel>,
+}
+
+impl QeIsvSvnVerifier {
+    fn new<'a, I>(tcb_levels: I) -> Self
+    where
+        I: IntoIterator<Item = &'a TcbLevel>,
+    {
+        Self {
+            tcb_levels: tcb_levels.into_iter().cloned().collect(),
+        }
+    }
+}
+
+impl<E: Accessor<IsvSvn>> Verifier<E> for QeIsvSvnVerifier {
+    type Value = (IsvSvn, Option<TcbLevel>);
+    fn verify(&self, evidence: &E) -> VerificationOutput<Self::Value> {
+        let isv_svn = evidence.get();
+        let tcb_levels = &self.tcb_levels;
+        let tcb_level = tcb_levels.iter().find_map(|tcb_level| {
+            if tcb_level.isv_svn().as_ref() <= isv_svn.as_ref() {
+                Some(tcb_level.clone())
+            } else {
+                None
+            }
+        });
+
+        let mut is_success_bool = false;
+        if let Some(tcb_level) = &tcb_level {
+            is_success_bool = tcb_level.status() == AdvisoryStatus::UpToDate
+                && tcb_level.advisory_ids().is_empty();
+        }
+        let is_success = if is_success_bool { 1 } else { 0 };
+
+        VerificationOutput::new((isv_svn, tcb_level), is_success.into())
+    }
+}
+
+impl Display for QeIsvSvnVerifier {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        write!(
+            f,
+            "The {} should correspond to an `UpToDate` level with no advisories, from: {:?}",
+            IsvSvn::spaced_struct_name(),
+            self.tcb_levels
+        )
+    }
+}
+
+impl VerificationMessage<(IsvSvn, Option<TcbLevel>)> for QeIsvSvnVerifier {
+    fn fmt_padded(
+        &self,
+        f: &mut Formatter<'_>,
+        pad: usize,
+        output: &VerificationOutput<(IsvSvn, Option<TcbLevel>)>,
+    ) -> core::fmt::Result {
+        let is_success = output.is_success();
+        let status = choice_to_status_message(is_success);
+        write!(f, "{:pad$}{status} {self}", "")?;
+        if (!is_success).into() {
+            let name = IsvSvn::spaced_struct_name();
+            let actual = &output.value;
+
+            write!(f, ", but the {name} of {} corresponds to ", actual.0)?;
+            if let Some(tcb_level) = &actual.1 {
+                write!(f, "{tcb_level:?}")?;
+            } else {
+                write!(f, "None")?;
+            }
```

Comment:
I went back and forth between this and
```rust
            if let Some(tcb_level) = &actual.1 {
                write!(f, ", but the {name} of {} corresponds to {tcb_level:?}", actual.0)?;
            } else {
                write!(f, ", but the {name} of {} corresponds to None", actual.0)?;
            }
```

I chose this one for DRY by default. While the alternative implementation in this comment allows for more tailoring of the "None" case, I think it's unlikely we'll see the "None" case practice as the qe identity seems to contain all the previous qe isv svn values.

Reply:
Nit: Is `write!(f, ", but the {name} of {} corresponds to {:?}", actual.0, actual.1)?;` not viable? It will include a `Some(..)` but not sure if that's an issue

Reply:
it probably isn't much of an issue. 
I was probably latched on to my previous solution too much.
Updated the code to print the debug version of the `Option<T>`